### PR TITLE
Enable service worker for legacy build

### PR DESCRIPTION
**Pull request by steverep**
## Proposed change

Create separate service workers for the modern and legacy builds.  The legacy SW targets browsers that also support them to avoid some large unnecessary transpiling and polyfills.  Adjusts the user-facing `__BUILD__` global to the more user-friendly "modern" and "legacy" terms.

Workbox is patched with https://github.com/GoogleChrome/workbox/pull/3234 in order to use the injection method which properly adjusts the source maps.

404 errors will occur without https://github.com/home-assistant/core/pull/120488.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Summary by CodeRabbit

- **New Features**
  - Added support for service workers in the legacy build profile.
  - Enhanced service worker registration with dynamic script paths based on build type.

- **Refactor**
  - Updated build configurations to distinguish between "modern" and "legacy" builds.
  - Refined service worker generation logic to handle modern and legacy builds separately.
  - Improved file handling for service worker scripts.

- **Bug Fixes**
  - Patched `workbox-build` to address specific issues.

- **Style**
  - Updated displayed JavaScript type from "latest" to "modern" in the UI.

- **Chores**
  - Removed unused `source-map-url` dependency.
  - Adjusted browser compatibility configurations.

### CODE REVIEW

**Comment by silamon** [comment]
If this is considered as a "breaking change", what is expected to be done by the user?

**Comment by steverep** [comment]
It's not breaking... I checked the wrong box.

File: build-scripts/bundle.cjs [hunk 1]
```diff
@@ -215,7 +220,13 @@ module.exports.config = {
     return {
       name: "frontend" + nameSuffix(latestBuild),
       entry: {
-        service_worker: "./src/entrypoints/service_worker.ts",
+        "service-worker":
+          !env.useRollup() && !latestBuild
+            ? {
+                import: "./src/entrypoints/service-worker.ts",
+                layer: "sw",
```
Comment (bramkragten) [inline build-scripts/bundle.cjs:227]: How do these layers work, can't find anything about it in the webpack docs...

Comment (steverep) [reply build-scripts/bundle.cjs:227]: Yeah the docs on it are quite sparse.  It's more or less a tag/label for a module.  You can assign it by entry like this, or [via a condition in the module config](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkslayer).  Then you can act on it [in cache groups](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkslayer), by issuer layer in the module config (like I do here to change the loader options), and they also get grouped in the stats output.

**Comment by frenck** [comment]
Core PR has been approved.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**